// @cubejs-backend/schema-compiler/dist/src/compiler/transpilers/CubePropContextTranspiler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CubePropContextTranspiler = exports.transpiledFields = exports.transpiledFieldsPatterns = void 0;
const t = __importStar(require("@babel/types"));
const ramda_1 = __importDefault(require("ramda"));
/* this list was generated by getTransformPatterns() with additional variants for snake_case */
exports.transpiledFieldsPatterns = [
    /\.sql$/,
    /sql$/,
    /(sqlTable|sql_table)$/,
    /^measures\.[_a-zA-Z][_a-zA-Z0-9]*\.(drillMemberReferences|drillMembers|drill_members)$/,
    /^measures\.[_a-zA-Z][_a-zA-Z0-9]*\.(orderBy|order_by)\.[0-9]+\.sql$/,
    /^measures\.[_a-zA-Z][_a-zA-Z0-9]*\.(timeShift|time_shift)\.[0-9]+\.(timeDimension|time_dimension)$/,
    /^measures\.[_a-zA-Z][_a-zA-Z0-9]*\.(reduceBy|reduce_by|groupBy|group_by|addGroupBy|add_group_by)$/,
    /^dimensions\.[_a-zA-Z][_a-zA-Z0-9]*\.(reduceBy|reduce_by|groupBy|group_by|addGroupBy|add_group_by)$/,
    /^(preAggregations|pre_aggregations)\.[_a-zA-Z][_a-zA-Z0-9]*\.indexes\.[_a-zA-Z][_a-zA-Z0-9]*\.columns$/,
    /^(preAggregations|pre_aggregations)\.[_a-zA-Z][_a-zA-Z0-9]*\.(timeDimensionReference|timeDimension|time_dimension|segments|dimensions|measures|rollups|segmentReferences|dimensionReferences|measureReferences|rollupReferences)$/,
    /^(preAggregations|pre_aggregations)\.[_a-zA-Z][_a-zA-Z0-9]*\.(timeDimensions|time_dimensions)\.\d+\.dimension$/,
    /^(preAggregations|pre_aggregations)\.[_a-zA-Z][_a-zA-Z0-9]*\.(outputColumnTypes|output_column_types)\.\d+\.member$/,
    /^contextMembers$/,
    /^includes$/,
    /^excludes$/,
    /^hierarchies\.[0-9]+\.levels$/,
    /^cubes\.[0-9]+\.(joinPath|join_path)$/,
];
exports.transpiledFields = new Set();
exports.transpiledFieldsPatterns?.forEach((r) => {
    const fields = r.toString().replace(/.*?([_a-zA-Z|][_a-zA-Z0-9|]*)([^_a-zA-Z0-9|]*)$/, '$1').split('|');
    fields.forEach((f) => exports.transpiledFields.add(f));
});
class CubePropContextTranspiler {
    constructor(cubeSymbols, cubeDictionary) {
        this.cubeSymbols = cubeSymbols;
        this.cubeDictionary = cubeDictionary;
    }
    traverseObject() {
        return {
            CallExpression: (path) => {
                if (t.isIdentifier(path.node.callee)) {
                    const args = path.get('arguments');
                    if (path.node.callee.name === 'cube' || path.node.callee.name === 'view') {
                        if (args?.[args.length - 1]) {
                            const cubeName = args[0].node.type === 'StringLiteral' && args[0].node.value ||
                                args[0].node.type === 'TemplateLiteral' &&
                                    args[0].node.quasis.length &&
                                    args[0].node.quasis[0].value.cooked;
                            args[args.length - 1].traverse(this.sqlAndReferencesFieldVisitor(cubeName));
                            args[args.length - 1].traverse(this.knownIdentifiersInjectVisitor('extends', name => this.cubeDictionary.resolveCube(name)));
                        }
                    }
                    else if (path.node.callee.name === 'context') {
                        args[args.length - 1].traverse(this.sqlAndReferencesFieldVisitor(null));
                    }
                }
            }
        };
    }
    transformObjectProperty(path, resolveSymbol) {
        CubePropContextTranspiler.replaceValueWithArrowFunction(resolveSymbol, path.get('value'));
    }
    static replaceValueWithArrowFunction(resolveSymbol, value) {
        const knownIds = CubePropContextTranspiler.collectKnownIdentifiers(resolveSymbol, value);
        value.replaceWith(t.arrowFunctionExpression(knownIds.map(i => t.identifier(i)), 
        // @todo Replace any with assert expression
        value.node, false));
    }
    sqlAndReferencesFieldVisitor(cubeName) {
        const resolveSymbol = n => this.cubeSymbols.resolveSymbol(cubeName, n) || this.cubeSymbols.isCurrentCube(n);
        return {
            ObjectProperty: (path) => {
                if (path.node.key.type === 'Identifier' && exports.transpiledFields.has(path.node.key.name)) {
                    const fullPath = this.fullPath(path);
                    // eslint-disable-next-line no-restricted-syntax
                    for (const p of exports.transpiledFieldsPatterns) {
                        if (fullPath.match(p)) {
                            this.transformObjectProperty(path, resolveSymbol);
                            return;
                        }
                    }
                }
            }
        };
    }
    fullPath(path) {
        // @ts-ignore
        let fp = path?.node?.key?.name || '';
        let pp = path?.parentPath;
        while (pp) {
            if (pp?.parentPath?.node?.type === 'ArrayExpression') {
                fp = `0.${fp}`;
                pp = pp?.parentPath;
                // @ts-ignore
            }
            else if (pp?.parentPath?.node?.key?.type === 'Identifier') {
                // @ts-ignore
                fp = `${pp?.parentPath?.node?.key?.name || '0'}.${fp}`;
                pp = pp?.parentPath?.parentPath;
            }
            else
                break;
        }
        return fp;
    }
    knownIdentifiersInjectVisitor(field, resolveSymbol) {
        return {
            ObjectProperty: (path) => {
                if (path.node.key.type === 'Identifier' && path.node.key.name.match(field)) {
                    this.transformObjectProperty(path, resolveSymbol);
                }
            }
        };
    }
    static collectKnownIdentifiers(resolveSymbol, path) {
        const identifiers = [];
        if (path.node.type === 'Identifier') {
            CubePropContextTranspiler.matchAndPushIdentifier(path, resolveSymbol, identifiers);
        }
        path.traverse({
            Identifier: (p) => {
                CubePropContextTranspiler.matchAndPushIdentifier(p, resolveSymbol, identifiers);
            }
        });
        return ramda_1.default.uniq(identifiers);
    }
    static matchAndPushIdentifier(path, resolveSymbol, identifiers) {
        if ((!path.parent ||
            (path.parent.type !== 'MemberExpression' || path.parent.type === 'MemberExpression' && path.key !== 'property')) &&
            resolveSymbol(path.node.name)) {
            identifiers.push(path.node.name);
        }
    }
}
exports.CubePropContextTranspiler = CubePropContextTranspiler;
