// @cubejs-backend/cubestore-driver/dist/codegen/http-query.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpQuery = void 0;
const flatbuffers = __importStar(require("flatbuffers"));
const http_table_js_1 = require("./http-table.js");
class HttpQuery {
    constructor() {
        this.bb = null;
        this.bb_pos = 0;
    }
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsHttpQuery(bb, obj) {
        return (obj || new HttpQuery()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsHttpQuery(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new HttpQuery()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    query(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    traceObj(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    inlineTables(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? (obj || new http_table_js_1.HttpTable()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    inlineTablesLength() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    static startHttpQuery(builder) {
        builder.startObject(3);
    }
    static addQuery(builder, queryOffset) {
        builder.addFieldOffset(0, queryOffset, 0);
    }
    static addTraceObj(builder, traceObjOffset) {
        builder.addFieldOffset(1, traceObjOffset, 0);
    }
    static addInlineTables(builder, inlineTablesOffset) {
        builder.addFieldOffset(2, inlineTablesOffset, 0);
    }
    static createInlineTablesVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startInlineTablesVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endHttpQuery(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createHttpQuery(builder, queryOffset, traceObjOffset, inlineTablesOffset) {
        HttpQuery.startHttpQuery(builder);
        HttpQuery.addQuery(builder, queryOffset);
        HttpQuery.addTraceObj(builder, traceObjOffset);
        HttpQuery.addInlineTables(builder, inlineTablesOffset);
        return HttpQuery.endHttpQuery(builder);
    }
}
exports.HttpQuery = HttpQuery;
