// @cubejs-backend/cubestore-driver/dist/codegen/http-result-set.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpResultSet = void 0;
const flatbuffers = __importStar(require("flatbuffers"));
const http_row_js_1 = require("./http-row.js");
class HttpResultSet {
    constructor() {
        this.bb = null;
        this.bb_pos = 0;
    }
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsHttpResultSet(bb, obj) {
        return (obj || new HttpResultSet()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsHttpResultSet(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new HttpResultSet()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    columns(index, optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__string(this.bb.__vector(this.bb_pos + offset) + index * 4, optionalEncoding) : null;
    }
    columnsLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    rows(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? (obj || new http_row_js_1.HttpRow()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    rowsLength() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    static startHttpResultSet(builder) {
        builder.startObject(2);
    }
    static addColumns(builder, columnsOffset) {
        builder.addFieldOffset(0, columnsOffset, 0);
    }
    static createColumnsVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startColumnsVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static addRows(builder, rowsOffset) {
        builder.addFieldOffset(1, rowsOffset, 0);
    }
    static createRowsVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startRowsVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endHttpResultSet(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createHttpResultSet(builder, columnsOffset, rowsOffset) {
        HttpResultSet.startHttpResultSet(builder);
        HttpResultSet.addColumns(builder, columnsOffset);
        HttpResultSet.addRows(builder, rowsOffset);
        return HttpResultSet.endHttpResultSet(builder);
    }
}
exports.HttpResultSet = HttpResultSet;
