// @cubejs-backend/schema-compiler/dist/src/scaffolding/ScaffoldingTemplate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScaffoldingTemplate = exports.SchemaFormat = void 0;
const formatters_1 = require("./formatters");
var SchemaFormat;
(function (SchemaFormat) {
    SchemaFormat["JavaScript"] = "js";
    SchemaFormat["Yaml"] = "yaml";
})(SchemaFormat || (exports.SchemaFormat = SchemaFormat = {}));
class ScaffoldingTemplate {
    constructor(dbSchema, driver, options = {
        snakeCase: false,
    }) {
        this.driver = driver;
        this.options = options;
        const formatterOptions = {
            snakeCase: Boolean(this.options.snakeCase),
            catalog: this.options.catalog,
        };
        this.formatStrategy =
            options.format === SchemaFormat.Yaml
                ? new formatters_1.YamlSchemaFormatter(dbSchema, this.driver, formatterOptions)
                : new formatters_1.JavaScriptSchemaFormatter(dbSchema, this.driver, formatterOptions);
    }
    generateFilesByTableNames(tableNames, schemaContext = {}) {
        return this.formatStrategy.generateFilesByTableNames(tableNames, schemaContext);
    }
    generateFilesByCubeDescriptors(cubeDescriptors, schemaContext = {}) {
        return this.formatStrategy.generateFilesByCubeDescriptors(cubeDescriptors, schemaContext);
    }
}
exports.ScaffoldingTemplate = ScaffoldingTemplate;
