// @cubejs-backend/api-gateway/dist/src/types/enums.js
"use strict";
/**
 * @license Apache-2.0
 * @copyright Cube Dev, Inc.
 * @fileoverview
 * API Gateway enums definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeGranularity = exports.ResultType = exports.QueryType = exports.OrderType = exports.MemberType = void 0;
/**
 * Query type enum.
 */
var QueryType;
(function (QueryType) {
    QueryType["REGULAR_QUERY"] = "regularQuery";
    QueryType["COMPARE_DATE_RANGE_QUERY"] = "compareDateRangeQuery";
    QueryType["BLENDING_QUERY"] = "blendingQuery";
})(QueryType || (exports.QueryType = QueryType = {}));
/**
 * Query result dataset formats enum.
 */
var ResultType;
(function (ResultType) {
    ResultType["DEFAULT"] = "default";
    ResultType["COMPACT"] = "compact";
})(ResultType || (exports.ResultType = ResultType = {}));
/**
 * Network query order types enum.
 */
var OrderType;
(function (OrderType) {
    OrderType["ASC"] = "asc";
    OrderType["DESC"] = "desc";
})(OrderType || (exports.OrderType = OrderType = {}));
/**
 * Query members types enum.
 */
var MemberType;
(function (MemberType) {
    MemberType["MEASURES"] = "measures";
    MemberType["DIMENSIONS"] = "dimensions";
    MemberType["SEGMENTS"] = "segments";
})(MemberType || (exports.MemberType = MemberType = {}));
/**
 * Time dimension granularity data type.
 */
var TimeGranularity;
(function (TimeGranularity) {
    TimeGranularity["QUARTER"] = "quarter";
    TimeGranularity["DAY"] = "day";
    TimeGranularity["MONTH"] = "month";
    TimeGranularity["YEAR"] = "year";
    TimeGranularity["WEEK"] = "week";
    TimeGranularity["HOUR"] = "hour";
    TimeGranularity["MINUTE"] = "minute";
    TimeGranularity["SECOND"] = "second";
})(TimeGranularity || (exports.TimeGranularity = TimeGranularity = {}));
