// @cubejs-backend/schema-compiler/dist/src/adapter/BaseSegment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseSegment = void 0;
class BaseSegment {
    constructor(query, segment) {
        this.query = query;
        this.segment = segment;
        this.isMemberExpression = false;
        if (segment.expression) {
            this.expression = segment.expression;
            this.expressionCubeName = segment.cubeName;
            // In case of SQL push down expressionName doesn't contain cube name. It's just a column name.
            this.expressionName = segment.expressionName || `${segment.cubeName}.${segment.name}`;
            this.isMemberExpression = !!segment.definition;
        }
    }
    filterToWhere() {
        return this.segmentSql();
    }
    segmentSql() {
        if (this.expression) {
            return this.query.evaluateSymbolSql(this.expressionCubeName, this.expressionName, this.definition(), 'segment');
        }
        return this.query.segmentSql(this);
    }
    filterParams() {
        return [];
    }
    segmentDefinition() {
        return this.query.cubeEvaluator.segmentByPath(this.segment);
    }
    isPostAggregate() {
        if (this.expression) { // TODO
            return false;
        }
        return this.definition().postAggregate;
    }
    definition() {
        if (this.expression) {
            return {
                sql: this.expression
            };
        }
        return this.segmentDefinition();
    }
    getMembers() {
        return [this];
    }
    cube() {
        if (this.expression) {
            return this.query.cubeEvaluator.cubeFromPath(this.expressionCubeName);
        }
        return this.query.cubeEvaluator.cubeFromPath(this.segment);
    }
    sqlDefinition() {
        return this.segmentDefinition().sql;
    }
    path() {
        if (this.expression) {
            return null;
        }
        return this.query.cubeEvaluator.parsePath('segments', this.segment);
    }
    expressionPath() {
        if (this.expression) {
            return `expr:${this.expression.expressionName}`;
        }
        return this.query.cubeEvaluator.pathFromArray(this.path());
    }
}
exports.BaseSegment = BaseSegment;
