// @cubejs-backend/schema-compiler/dist/src/adapter/BaseGroupFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseGroupFilter = void 0;
class BaseGroupFilter {
    constructor(filter) {
        this.values = filter.values;
        this.operator = filter.operator;
        this.measure = filter.measure;
        this.dimension = filter.dimension;
    }
    isDateOperator() {
        return false;
    }
    conditionSql(column) {
        return `(\n${this.values.map(f => f.conditionSql(column)).join(` ${this.operator.toUpperCase()} `)}\n)`;
    }
    filterToWhere() {
        const r = this.values.map(f => {
            const sql = f.filterToWhere();
            if (!sql) {
                return null;
            }
            return `(${sql})`;
        }).filter(x => x).join(` ${this.operator.toUpperCase()} `);
        if (!r.length) {
            return null;
        }
        return r;
    }
    filterParams() {
        return this.values.map(f => f.filterParams());
    }
    getMembers() {
        return this.values.map(f => {
            if (f.getMembers) {
                return f.getMembers();
            }
            return f;
        });
    }
}
exports.BaseGroupFilter = BaseGroupFilter;
